### ISSUE TITLE
Migrate k8s-infra-prow-build to a nodepool with more IOPS

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/k8s.io#1187]
This is followup to https://github.com/kubernetes/k8s.io/issues/1168 and https://github.com/kubernetes/k8s.io/issues/1173, made possible by quota changes done via https://github.com/kubernetes/k8s.io/issues/1132

My goal is to make these graphs go down
![Screen Shot 2020-08-28 at 10 24 01 AM](https://user-images.githubusercontent.com/49258/91595625-b138d480-e918-11ea-914a-cd1570be3207.png)

Our jobs are hitting I/O limits (both IOPS and throughput).  That was made extra clear last weekend when we switched to larger nodes, thus causing more jobs to share the same amount of I/O.  

We're seeing more jobs scheduled into the cluster now that v1.20 PRs are being merged.  While our worst case node performance is about the same, we are seeing more throttling across the cluster in aggregate.

Kubernetes doesn't give us a way to provision I/O, so we're left optimizing per-node performance.  Based on https://cloud.google.com/compute/docs/disks/performance I think we can get just under 2x the IOPS for a ~14% increase in cluster cost.

From there, going to next tier would require a 90% increase in cost, for only 66% more performance.

The most ideal thing would be local SSD but:
- [that's not a node bootdisk option at present](https://github.com/kubernetes/kubernetes/issues/76553)
- [cluster autoscaling doesn't work with local PersistentVolumes](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/local-ssd#limitations)
- leaving us to maybe cobble something together that would have to replace `emptyDir` volumes with `hostPath` volumes, which sounds like a maintenance nightmare

**Comment by spiffxp** [comment]
Opened https://github.com/kubernetes/k8s.io/pull/1186 to start migrating to first option (14% more cost for ~100% more IOPS)

**Comment by spiffxp** [comment]
New nodepool is up, old nodepool cordoned
```
spiffxp@cloudshell:~ (k8s-infra-prow-build)$ for n in $(k get nodes -l cloud.google.com/gke-nodepool=pool3-20200824192452986800000001 -o=name); do k cordon $n; done
node/gke-prow-build-pool3-2020082419245298-65156f0e-0cwp cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-4xr9 cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-7bnv cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-7x2j cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-9q3g cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-fmpb cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-v5l5 cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-z4lq cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-5sqt cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-6kr5 cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-7wr8 cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-b5ns cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-h4v9 cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-nb31 cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-0sn6 cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-8qq0 cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-b496 cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-jt7h cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-lk1n cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-lnkj cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-mrt2 cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-qwl2 cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-s9x6 cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-tkwx cordoned
node/gke-prow-build-pool3-2020082419245298-f02ec66b-x2g2 cordoned
```

**Comment by spiffxp** [comment]
And I forgot to disable autoscaling for pool3 until just now
```
spiffxp@cloudshell:~ (k8s-infra-prow-build)$ for n in $(k get nodes -l cloud.google.com/gke-nodepool=pool3-20200824192452986800000001 -o=name); do k cordon $n; done | grep -v already
node/gke-prow-build-pool3-2020082419245298-65156f0e-4bsj cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-bzg3 cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-l8n1 cordoned
node/gke-prow-build-pool3-2020082419245298-65156f0e-n6mz cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-1bgk cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-968j cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-fpqb cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-nv8r cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-qgx2 cordoned
node/gke-prow-build-pool3-2020082419245298-d2051b67-s4kj cordoned
```

**Comment by spiffxp** [comment]
Deleted boskos
```
$ date; k delete pod -n test-pods -l app=boskos; date
Fri 28 Aug 2020 08:01:08 PM UTC
pod "boskos-564f5594dd-sk2jv" deleted
Fri 28 Aug 2020 08:01:17 PM UTC

$ date; k delete pod -n test-pods -l app=boskos-janitor; date
Fri 28 Aug 2020 08:01:38 PM UTC
pod "boskos-janitor-58c6d75dc9-684lm" deleted
pod "boskos-janitor-58c6d75dc9-s9b6s" deleted
Fri 28 Aug 2020 08:06:47 PM UTC

$ date; k delete pod -n test-pods -l app=boskos-reaper; date
Fri 28 Aug 2020 08:16:51 PM UTC
pod "boskos-reaper-56b467f9d8-c6rfw" deleted
Fri 28 Aug 2020 08:16:57 PM UTC
```
Waiting on the following to finish up
```
$ date; k get pods -n test-pods --field-selector=status.phase=Running -o=json | jq -r '.items | map(select(.spec.nodeName | match("pool3")))[] | "\(.status.startTime) \(.metadata.labels["prow.k8s.io/job"]) \(.metadata.name) \(.spec.nodeName)"' | sort | tee old-nodepool-jobs
Fri 28 Aug 2020 08:43:02 PM UTC
2020-08-28T16:05:56Z ci-kubernetes-e2e-gce-cos-k8sbeta-serial 53dbcab4-e948-11ea-aa1a-c6580c04344b gke-prow-build-pool3-2020082419245298-f02ec66b-0sn6
2020-08-28T16:12:56Z ci-kubernetes-e2e-gci-gce-serial 4e00d2eb-e949-11ea-aa1a-c6580c04344b gke-prow-build-pool3-2020082419245298-f02ec66b-tkwx
2020-08-28T17:34:56Z ci-kubernetes-e2e-gce-cos-k8sstable1-serial c2d6b387-e954-11ea-aa1a-c6580c04344b gke-prow-build-pool3-2020082419245298-d2051b67-b5ns
2020-08-28T18:26:56Z ci-kubernetes-gce-conformance-latest-1-16 067d54f8-e95c-11ea-aa1a-c6580c04344b gke-prow-build-pool3-2020082419245298-f02ec66b-0sn6
2020-08-28T18:47:48Z ci-kubernetes-gce-conformance-latest-kubetest2 8a2d1fdd-e95e-11ea-aa1a-c6580c04344b gke-prow-build-pool3-2020082419245298-65156f0e-bzg3
```

**Comment by spiffxp** [comment]
Trimmed empty nodes
![Screen Shot 2020-08-28 at 1 46 55 PM](https://user-images.githubusercontent.com/49258/91613738-f7039600-e934-11ea-8142-72ba0df3ad9a.png)

**Comment by spiffxp** [comment]
Removed old nodepool with https://github.com/kubernetes/k8s.io/pull/1188

**Comment by spiffxp** [comment]
Holding this open to see what impact, if any, this has on the graphs shown in the description

https://console.cloud.google.com/monitoring/dashboards/custom/10925237040785467832?project=k8s-infra-prow-build&timeDomain=1w

**Comment by spiffxp** [comment]
No real change in the graphs, other than a reflection of PR traffic.  This certainly didn't make things _worse_ and isn't urgently more expensive, so I'm not inclined to rollback at the moment.

![Screen Shot 2020-08-31 at 12 53 36 PM](https://user-images.githubusercontent.com/49258/91760981-109d1b80-eb89-11ea-8ab7-868898c2c371.png)

Supposedly PER_GB means increase the persistent disk size (ref: https://cloud.google.com/compute/docs/disks/review-disk-metrics#throttling_metrics)

One option would be to increase disk size to the next "tier" and see what happens.  But I think I'd like to do a little more reading and focused testing to understand what's going on, and what options we have.

**Comment by spiffxp** [comment]
/remove-lifecycle rotten
Creating node pools with local boot SSD's may be possible in a later version, but we're not there yet

**Comment by spiffxp** [comment]
I'm really interested in seeing this happen, but I can't guarantee I'll have bandwidth this cycle, so leaving out of milestone.  Gated on migrating cluster to 1.18

**Comment by spiffxp** [comment]
/priority backlog
/milestone v1.21
changed my mind about milestoning, I'll assign low priority

**Comment by spiffxp** [comment]
>Note: Local SSDs require machine type n1-standard-1 or larger; the default machine type, e2-medium is not supported. You can learn more about machine types in the Compute Engine documentation.

This would be a reason to keep the cluster using n1 nodes instead of switching to e2 nodes for lower cost.

**Comment by spiffxp** [comment]
https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/local-ssd#creating_a_node_pool_using_ephemeral_storage_on_local_ssds

Last I checked, the terraform provider we use didn't yet have support for `--ephemeral-storage local-ssd-count=number-of-disks`

Our options are:
- check again, maybe it's been added
- open an issue or PR to get support added
- see if it's possible to  manually provision with `gcloud` and then manage with terraform thereafter with `terraform import`

**Comment by BenTheElder** [comment]
looks like ephemeral_storage_config local_ssd_count has landed
https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/container_cluster#ephemeral_storage_config

**Comment by ameukam** [comment]
Unblocked by #2017.

**Comment by BenTheElder** [comment]
We should start with a tainted + labeled pool so we can just schedule certain jobs to it to try and then we can see if we want to migrate everything to it.

**Comment by aojea** [comment]
> We should start with a tainted + labeled pool so we can just schedule certain jobs to it to try and then we can see if we want to migrate everything to it.

we have some canary jobs that can be good candidates

**Comment by BenTheElder** [comment]
I would like to move the kubernetes kind presubmits to it. I don't expect negative outcomes from this nodepool, but if the outcomes are not sufficiently positive we might not bother moving everything (it also costs more ...)

**Comment by BenTheElder** [comment]
/help
anyone should be able to PR the terraform and job configs, I can try to help understand these but admittedly haven't just done it myself yet because too many things keep coming up ...

**Comment by spiffxp** [comment]
Provisioning a nodepool with `ephemeral-ssd-storage=true` as a taint: https://github.com/kubernetes/k8s.io/pull/2835

I'll try cutting over some canary presubmits to see what the behavior is

**Comment by spiffxp** [comment]
https://github.com/kubernetes/test-infra/pull/23783 will cutover:
- pull-kubernetes-e2e-gce-canary
- pull-kubernetes-integration-go-canary
- pull-kubernetes-e2e-kind-canary
- pull-kubernetes-unit-go-canary
- pull-kubernetes-verify-go-canary

Which are all manually triggered

**Comment by spiffxp** [comment]
After an initial round of canary jobs against a single PR, I have kicked off the canary jobs against a handful of arbitrary kubernetes/kubernetes PR's to trigger autoscaling and evaluate the node disk usage under some level of concurrency / load

https://console.cloud.google.com/monitoring/dashboards/builder/f0163540-a8b7-4618-8308-66652d3d4794?project=k8s-infra-prow-build&dashboardBuilderState=%257B%2522editModeEnabled%2522:false%257D&timeDomain=1h is the dashboard I'm using to watch the pot boil

Old pool is on left (pool4), new pool is on right (pool5).  By default, Google Cloud Monitoring doesn't appear to let me manually set the Y axis scales.  I added an arbitrary threshold to each graph to give them the same scale.  You can see we're experiencing way less throttling with the new pool.

![Screen Shot 2021-09-27 at 4 32 59 PM](https://user-images.githubusercontent.com/49258/134999046-d09965d0-3bd0-488e-99f0-cc8b68089667.png)

**Comment by spiffxp** [comment]
Need to cost out and estimate quota before rolling this out more generally.  The numbers look good enough that I'm interested in doing so.  However...

I legitimately can't tell that there's any immediately obvious speedup from doing this?  I'll let the other jobs finish and take a look at PR history for a quick check tomorrow.  The only other thing I can think that this might allow is lowering of some CPU/memory resource limits to pack jobs more densely if they're in fact not going to be as noisy to each other.  That will probably require more attention than I have time for right now.

**Comment by spiffxp** [comment]
https://cloud.google.com/compute/disks-image-pricing#localssdpricing

Local SSDs are $30/mo, so x2 = 60/mo
Zonal SSDs are .17/GB/mo, so 500GB = 85/mo
Zonal PDs are .04/GB/mo, so 100GB = 4/mo

https://cloud.google.com/compute/vm-instance-pricing

n1-highmem-8 are ~ 241/mo

pool4 instances are n1-highmem-8 + 500GB pd-ssd = 241 + 85 = 326
pool5 instances are n1-highmem-8 + 100GB pd-standard + 2x local SSD = 241 + 60 + 4 = 305

That's about 7% savings... we could bring that to 16% if we used only 1 local SSD.  Looking at our total spend just for k8s-infra-prow-build over the last year, it was ~258K.  7% savings would be ~18K, 16% ~40K.  Not nothing, but not incredibly significant compared against our total budget.

Quota: 
- pool4 is capped at 80 instances * 3 zones * 2 local ssds * 375GB/ssd = 180,000 GB
- current quota for us-central1 is 100,000 GB

Conclusions:
- bump quota to 200,000 GB local SSD
- pool5's current configuration doesn't appear to impact performance negatively
- pool5 will be slightly cheaper than before
- let's move forward with this, and consider denser/higher-CPU nodes in the future

**Comment by spiffxp** [comment]
Looks like I'm going to have to force recreation of the node pool to drop the taint
```
$ gcloud beta container node-pools update pool5-20210927201718330300000001 --project=k8s-infra-prow-build --zone=us-central1 --cluster=prow-build --node-taints=""
ERROR: (gcloud.beta.container.node-pools.update) ResponseError: code=400, message=Updates for 'taints' are not supported in node pools with autoscaling enabled (as a workaround, consider temporarily disabling autoscaling or recreating the node pool with the updated values.).
```

**Comment by aojea** [comment]
out of curiosity, why 2 local ssds?

**Comment by spiffxp** [comment]
I was seeking to:
- match to 500GB amount we used previously, and 375GB was too little
- over-provision capacity so as not to saturate a single ssd

**Comment by spiffxp** [comment]
OK, migration to new nodepool with local SSDs for ephemeral storage complete, see https://github.com/kubernetes/k8s.io/pull/2839 for details

Throttled I/O is way down post-migration
![Screen Shot 2021-09-28 at 10 22 43 AM](https://user-images.githubusercontent.com/49258/135135483-4b6298ec-1a59-419e-8a92-ed2ded16f26d.png)

Throttled bytes from old pool nodes on the left, new pool nodes on the right
![Screen Shot 2021-09-28 at 10 23 53 AM](https://user-images.githubusercontent.com/49258/135135803-f336ffef-90b8-4737-8369-243bd87b7e09.png)

**Comment by spiffxp** [comment]
I'll hold this open for a day to see if this had any negative impact, but I otherwise consider this issue closed.

It'll take a bit to determine whether this has had any impact on job / build time.  Again, my guess based on a brief survey of the canary jobs from yesterday is negligible impact at best, but hopefully fewer noisy neighbors.  We lack a great way to display this data at present, though I suspect the data will be available in some form in bigquery `k8s-gubernator:build`

**Comment by spiffxp** [comment]
/close
Calling this done
